System user, local RPC, demos use RPC

exFalso opened the pull request:
This is a reopen from bitbucket. To sum up:

This PR is the first step to factoring out of the webserver into a separate process. This requires plugin web apis to not depend on the internal ServiceHub, but instead use the RPC interface. As web auth is not yet figured out a god-mode user was added for the webserver.

It also required some code moving, most importantly the RPC interface was moved from `:node` to `:core`

shamsasari commented on node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt:
```diff
@@ -84,8 +91,41 @@ class CordaRPCOpsImpl(
+        fun stateMachineInfoFromFlowStateMachineImpl(stateMachine: FlowStateMachineImpl<*>): StateMachineInfo {
+            return StateMachineInfo(
+                    id = stateMachine.id,
+                    flowLogicClassName = stateMachine.logic.javaClass.simpleName,
```
Perhaps document this property that it's not the fully-qualified class name but rather the class simple name.

rick-r3 replied:
Will we regret this not being fully qualified later?  The client can always display the truncated name if we wish but hard to add back the package name.

exFalso replied:
I think the only place this is used currently is the explorer. I'll change it to be fully qualified. @patrickkuo the ui code may need some adjustment

shamsasari commented on core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt:
```diff
@@ -160,6 +143,6 @@ inline fun <T : Any, A, B, C, D, reified R : FlowLogic<T>> CordaRPCOps.startFlow
         val id: StateMachineRunId,
-        val progress: Observable<ProgressTracker.Change>,
+        val progress: Observable<String>,
         val returnValue: Observable<A>
```
As a side point, I disagree that this should be an Observable when we know it will emit only one value. ListenableFuture better captures this intent. WDYT?

mikehearn replied:
I think the reason is that the RPC system doesn't know how to transport a ListenableFuture, whereas it does know how to transport observables.

rick-r3 replied:
Wouldn't hurt to put a comment setting expectations of a single observation here.  Generally some docs for ``FlowHandle`` would be good.

exFalso replied:
Exactly. We may be able to extend it with some additional functionality, or perhaps even expose a lower level api, but for now only Observables are supported, adding support for something else is out of scope for this pr

mikehearn replied:
Yes - we could extend RPC to support marshalling of futures, but it's a separate project. For now a comment explaining this would be sufficient.

shamsasari commented on core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt:
```diff
@@ -160,6 +143,6 @@ inline fun <T : Any, A, B, C, D, reified R : FlowLogic<T>> CordaRPCOps.startFlow
 data class FlowHandle<A>(
         val id: StateMachineRunId,
-        val progress: Observable<ProgressTracker.Change>,
+        val progress: Observable<String>,
```
This was a nice type-safe `Observable` of `Change`s. Why the change?

exFalso replied:
It never worked as the serialisers are not registered, we just didn't notice. And trying to register the serialisers is hacky as the ProgressTracker code has some hidden classes.

exFalso replied:
Correction: the issue wasn't hidden classes, but other nested Observables and backlinks. The `String`s are also consistent with the other way of retrieving progress tracker events (through the state machine rpc). We'll need to rethink the progress tracker code anyway, perpahs then we can think about a more serialisation-friendly data model

shamsasari commented on node/src/main/kotlin/net/corda/node/services/RPCUserService.kt:
```diff
@@ -11,22 +13,44 @@ import net.corda.node.services.config.FullNodeConfiguration
     val users: List<User>
+
+    /**
+     * TODO The system user is used by the web api until we implement web auth properly
```
Can you update this doc to make it explicit that this is a temporary hack.

shamsasari commented on node/src/main/kotlin/net/corda/node/services/RPCUserService.kt:
```diff
@@ -11,22 +13,44 @@ import net.corda.node.services.config.FullNodeConfiguration
+    fun addSystemUserPermission(permission: String)
+    companion object {
+        val systemUserUsername = "~systemUser"
+        val systemUserPassword: String = SecureHash.randomSHA256().toString()
```
Explicit type isn't needed

exFalso replied:
IntelliJ gives the usual warning about platform types if it's not specified

shamsasari replied:
Ah, `SecureHash.toString` doesn't explicitly specify the return type (because it's calling Java code). Fixing that will remove the need here.

mikehearn replied:
Isn't this redundant after your change to use TLS certs for intra-node components, shams?

shamsasari replied:
Good spot! Have the webserver login to the Artemis server with the node user and because it's using the same keystore it will use the same TLS cert and get logged in as the node and get full permissions. The only thing missing would be to give full RPC permissions to the node user. I think that can easily fixed by updating `requirePermission`. I like it!

@exFalso my branch is shams-p2p-auth.

exFalso replied:
Ok, that sounds much cleaner, i'll rebase onto your branch

shamsasari commented on node/src/main/kotlin/net/corda/node/services/RPCUserService.kt:
```diff
@@ -11,22 +13,44 @@ import net.corda.node.services.config.FullNodeConfiguration
+    private val _users: HashMap<String, User>
+
+    init {
+        val defaultUsers = listOf(User(RPCUserService.systemUserUsername, RPCUserService.systemUserPassword, setOf()))
```
Can you statically import the these two - the line looks a little cluttered

shamsasari commented on node/src/main/kotlin/net/corda/node/services/messaging/RPCStructures.kt:
```diff
@@ -92,7 +78,7 @@ val CURRENT_RPC_USER: ThreadLocal<User> = ThreadLocal()
 fun requirePermission(permission: String) {
     if (permission !in CURRENT_RPC_USER.get().permissions) {
-        throw PermissionException("User not permissioned for $permission")
+        throw PermissionException("User not permissioned for $permission, permissions are ${CURRENT_RPC_USER.get().permissions}")
```
Extract out the current `User` into a local variable

shamsasari commented on samples/irs-demo/src/main/kotlin/net/corda/irs/plugin/IRSPlugin.kt:
```diff
@@ -2,25 +2,26 @@ package net.corda.irs.plugin
             "irsdemo" to javaClass.classLoader.getResource("irsweb").toExternalForm()
     )
-    override val servicePlugins: List<Class<*>> = listOf(FixingFlow.Service::class.java)
+    override val servicePlugins = listOf({ services: PluginServiceHub -> FixingFlow.Service(services) })
```
`FixingFlow::Service` should work here right?

shamsasari commented on samples/raft-notary-demo/src/main/kotlin/net/corda/notarydemo/flows/DummyIssueAndMove.kt:
```diff
@@ -0,0 +1,31 @@
+import net.corda.core.node.recordTransactions
+import net.corda.core.transactions.SignedTransaction
+import java.util.*
+
```
Remove extra blank line

shamsasari commented on samples/simm-valuation-demo/src/main/kotlin/net/corda/vega/services/SimmService.kt:
```diff
@@ -17,14 +18,15 @@ import java.time.LocalDate
-                IRSTradeFlow.Service::class.java)
         override val staticServeDirs: Map<String, String> = mapOf("simmvaluationdemo" to javaClass.classLoader.getResource("simmvaluationweb").toExternalForm())
+        override val servicePlugins: List<(PluginServiceHub) -> Any> = listOf(
+                { services -> SimmFlow.Service(services) },
```
`SimmFlow::Service` and similarly below

shamsasari commented on samples/trader-demo/src/main/kotlin/net/corda/traderdemo/plugin/TraderDemoPlugin.kt:
```diff
@@ -2,17 +2,18 @@ package net.corda.traderdemo.plugin
             SellerFlow::class.java.name to setOf(Party::class.java.name, Amount::class.java.name)
     )
-    override val servicePlugins: List<Class<*>> = listOf(BuyerFlow.Service::class.java)
+    override val servicePlugins = listOf({ services: PluginServiceHub -> BuyerFlow.Service(services) })
```
And here

mikehearn commented:
You might want to rebase on Shams' work, which already introduces various users for node components.

rick-r3 commented on core/sandbox/src/main/java/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,3 @@
+Manifest-Version: 1.0
+Main-Class: com.r3cev.NonDeter
```
Is this file intentionally here?

rick-r3 commented on core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt:
```diff
@@ -112,6 +84,17 @@ interface CordaRPCOps : RPCOps {
+
+    fun attachmentExists(id: SecureHash): Boolean
+    fun uploadAttachment(jar: InputStream): SecureHash
+    fun nodeTimeInUTC(): LocalDateTime
```
Good idea.  And convert in the code using the RPC client if we need to.

rick-r3 commented:
It looks like some rogue files have got swept up here, including some zip'd log files.  Please double check and remove them.

exFalso commented:
Addressed all comments, ball in your courts

mikehearn commented:
I have a half-finished review, hold on.

mikehearn commented on node/src/main/kotlin/net/corda/node/internal/Node.kt:
```diff
@@ -121,6 +124,17 @@ class Node(override val configuration: FullNodeConfiguration, networkMapAddress:
         userService = RPCUserServiceImpl(configuration)
+
+        // Add System user permissions
+        defaultFlowWhiteList.forEach {
```
I think this will conflict with Shams' work - might want to rebase onto his.

mikehearn commented on node/src/main/kotlin/net/corda/node/internal/Node.kt:
```diff
@@ -121,6 +124,17 @@ class Node(override val configuration: FullNodeConfiguration, networkMapAddress:
+        defaultFlowWhiteList.forEach {
+            userService.addSystemUserPermission(startFlowPermission(it.key))
+        }
+        pluginRegistries.forEach {
```
I know forEach can be convenient but I'm not a huge fan of the scope shadowing of `it` here. A good old fashioned for loop is basically a foreach in kotlin ...

mikehearn commented on node/src/main/kotlin/net/corda/node/internal/Node.kt:
```diff
@@ -297,13 +306,20 @@ class Node(override val configuration: FullNodeConfiguration, networkMapAddress:
         super.start()
         // Only start the service API requests once the network map registration is complete
-        networkMapRegistrationFuture.then {
+        thread(name = "WebServer") {
```
Why put this into the background? I think it will mess up the measurement of startup time at least, but it seems an unrelated change?

mikehearn commented on node/src/main/kotlin/net/corda/node/services/messaging/RPCStructures.kt:
```diff
@@ -218,6 +205,9 @@ private class RPCKryo(observableSerializer: Serializer<Observable<Any>>? = null)
         register(FlowHandle::class.java)
         register(KryoException::class.java)
         register(StringBuffer::class.java)
+        register(Unit::class.java)
```
Are we really serialising Unit over the wire?!

exFalso replied:
Yes, there are flows that return Unit

mikehearn commented on samples/raft-notary-demo/src/main/kotlin/net/corda/notarydemo/api/NotaryDemoApi.kt:
```diff
@@ -3,28 +3,29 @@ package net.corda.notarydemo.api
+class NotaryDemoApi(val rpc: CordaRPCOps) {
     private val notary by lazy {
-        services.networkMapCache.getAnyNotary() ?: throw IllegalStateException("No notary found on the network")
+        rpc.networkMapUpdates().first.first { it.advertisedServices.any { it.info.type.isNotary() } }.notaryIdentity
```
This sort of construct makes me think the rpc framework should support a notion of idempotent returns (i.e. an RPC always returns the same thing and can be de-duplicated on the client side).

exFalso replied:
You mean rpc memoization? Makes sense, but requires some thought e.g. it'd mean that from the first call on we'll hold a reference to the returned observable.

mikehearn commented on samples/trader-demo/src/main/kotlin/net/corda/traderdemo/api/TraderDemoApi.kt:
```diff
@@ -29,34 +34,44 @@ class TraderDemoApi(val services: ServiceHub) {
-        services.fillWithSomeTestCash(params.amount.DOLLARS,
-                outputNotary = notary,
-                ownedBy = services.myInfo.legalIdentity.owningKey)
+        val notary = rpc.networkMapUpdates().first.first { it.legalIdentity.name == params.notary }
```
This is such an ugly pattern .... it implies to me there should be a dedicated RPC for this common need.

exFalso commented on core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt:
```diff
@@ -112,6 +84,17 @@ interface CordaRPCOps : RPCOps {
     fun getVaultTransactionNotes(txnId: SecureHash): Iterable<String>
+
+    fun attachmentExists(id: SecureHash): Boolean
+    fun uploadAttachment(jar: InputStream): SecureHash
```
Yes, I was thinking we may need to expose some lower level api to stream efficiently.

I was also surprised, but it just worked out of the box

exFalso commented on core/src/main/kotlin/net/corda/core/node/CordaPluginRegistry.kt:
```diff
@@ -1,52 +1,52 @@
-     * Cordapp code can be loaded dynamically.
-     */
-    open val webApis: List<Class<*>> = emptyList()
+abstract class CordaPluginRegistry(
```
What do you mean? These used to be `val`s, I kept that, just moved them from the body of the class to it's constructor

exFalso replied:
Wrt lambdas, the api used to be `SimmFlow.Service::class.java`, this changed to `SimmFlow::Service`. The difference is that this does compile time checks, whereas the previous version did it with reflection, finding the first constructor that seemed good.

exFalso commented on core/src/main/kotlin/net/corda/flows/FinalityFlow.kt:
```diff
@@ -16,7 +16,9 @@ import net.corda.core.utilities.ProgressTracker
                    val participants: Set<Party>,
-                   override val progressTracker: ProgressTracker = tracker()) : FlowLogic<Unit>() {
+                   override val progressTracker: ProgressTracker) : FlowLogic<Unit>() {
+    constructor(transaction: SignedTransaction, participants: Set<Party>) : this(transaction, participants, tracker())
```
Yes i tried that but it's not enough:S. Perhaps worth an issue raise @ jetbrains?

shamsasari commented on core/src/main/kotlin/net/corda/core/node/CordaPluginRegistry.kt:
```diff
@@ -1,52 +1,52 @@
+         * List of lambdas returning JAX-RS objects. They may only depend on the RPC interface, as the webserver should
+         * potentially be able to live in a process separate from the node itself.
+         */
+        open val webApis: List<(CordaRPCOps) -> Any> = emptyList(),
```
They are Java compatible but just that they are not the standard lambda interfaces that Java 8 uses.

exFalso commented on node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt:
```diff
@@ -84,8 +92,41 @@ class CordaRPCOpsImpl(
+        fun stateMachineUpdateFromStateMachineChange(change: StateMachineManager.Change): StateMachineUpdate {
+            return when (change.addOrRemove) {
+                AddOrRemove.ADD -> {
+                    val stateMachineInfo = StateMachineInfo(
```
No, this is based on a state machine change, whereas the other is based on a statemachineimpl. I think the only code shared is `(..).logic.javaClass.name` and `(..).logic.track()`

mikehearn commented:
Is this ball in my court? I think it needs a rebase.

exFalso commented:
This is now rebased, comments addressed. The NODE user handling is quite ugly, added TODOs to remove those pieces of code once the webserver doesnt need it

shamsasari commented on core/src/main/kotlin/net/corda/core/serialization/Kryo.kt:
```diff
@@ -202,6 +200,55 @@ class ImmutableClassSerializer<T : Any>(val klass: KClass<T>) : Serializer<T>()
+        val buffer = ByteArray(4096)
+        while (true) {
+            val numberOfBytesRead = stream.read(buffer)
+            if (numberOfBytesRead > 0) {
```
-1 represents end of stream

exFalso replied:
Yes

shamsasari replied:
This will terminate on reading 0 bytes but that's not what the docs say. It only ends on -1.

exFalso replied:
It can only return 0 if the input buffer is of size 0, which will not happen.
I'll just change it to `== -1`

shamsasari commented on core/src/main/kotlin/net/corda/core/serialization/Kryo.kt:
```diff
@@ -202,6 +200,55 @@ class ImmutableClassSerializer<T : Any>(val klass: KClass<T>) : Serializer<T>()
 
+// TODO This is a temporary inefficient serialiser for sending InputStreams through RPC. This may be done much more
+// efficiently using Artemis's large message feature.
+class InputStreamSerializer : Serializer<InputStream>() {
```
object

shamsasari commented on core/src/main/kotlin/net/corda/core/serialization/Kryo.kt:
```diff
@@ -202,6 +200,55 @@ class ImmutableClassSerializer<T : Any>(val klass: KClass<T>) : Serializer<T>()
+            }
+        }
+
+        return object : InputStream() {
```
Why not flatten `chunks` and then use a `ByteArrayInputStream`?

shamsasari commented on node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt:
```diff
@@ -184,7 +184,9 @@ class ArtemisMessagingServer(override val config: NodeConfiguration,
         securityRoles[RPC_REQUESTS_QUEUE] = setOf(nodeInternalRole, restrictedRole(RPC_ROLE, send = true))
-        for ((username) in userService.users) {
+        // TODO remove NODE_USER once webserver doesn't need it
+        val possibleClientUserNames = userService.users.map { it.username } + listOf(NODE_USER)
```
Make it explicit for the NODE USER: `securityRoles["$CLIENTS_PREFIX$NODE_USER.rpc.*"] = setOf(nodeInternalRole)`

shamsasari commented on node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt:
```diff
@@ -184,7 +184,10 @@ class ArtemisMessagingServer(override val config: NodeConfiguration,
-        for ((username) in userService.users) {
+        // TODO remove the NODE_USER role once the webserver doesn't need it
+        securityRoles["$CLIENTS_PREFIX$NODE_USER.rpc.*"] = setOf(nodeInternalRole)
+        val possibleClientUserNames = userService.users.map { it.username }
```
Revert this change

shamsasari commented on node/src/main/kotlin/net/corda/node/services/messaging/RPCDispatcher.kt:
```diff
@@ -153,9 +160,19 @@ abstract class RPCDispatcher(val ops: RPCOps, val userService: RPCUserService) {
     }
 
+    // TODO remove this User once webserver doesn't need it
+    val nodeUser = User(NODE_USER, NODE_USER, setOf())
```
You don't need to expose this as you're not using the object elsewhere.

exFalso commented:
Comments addressed, @mikehearn is this good to go?

mikehearn commented:
Yes, that looks OK now. The InputStream serialiser still looks weird - why bother chunking when you read the whole thing into RAM anyway - but as we'll toss it sooner or later it doesn't matter much.

The node internal user is not ideal: it'd be better to do least privilege for node internal components too, but that can be tackled as part of later security work. You could maybe add a brief discussion where NODE_USER is defined on the topic of privilege minimisation for node components. That'd be nice to have, to show readers of the code that we're thinking about it.
